pull_github_pr.sh: add notes to commit with reference to PR

yaronkaikov opened the pull request:
Today when we need to merge a PR with one commit, we add the `closes` prefix and then merge it using the `pull_github_pr.sh` script. This is causing a difference in commit sha between the one we see in the PR and the promoted commit. Causing our automation bot (Mergify) to backport wrong commit SHA.

This change will use `git notes` instead to update the reference to the PR.

Also instead of using `closes` which is GitHub closing keyword, let's use `From pull request:`

denesb commented on scripts/pull_github_pr.sh:
```diff
@@ -9,33 +9,18 @@
 fi
 
-for required in jq curl; do
+for required in jq gh; do
```
Do we need extra steps to authenticate with `gh`, or will the previous way with `GITHUB_LOGIN` and `GITHUB_TOKEN` keep working?

yaronkaikov replied:
if `GITHUB_TOKEN` is available it will be used. other option are available as explained below
```
[yaronkaikov@london]~/git/test-pkg (fix_next_with_noARM)$ gh auth login --help
Authenticate with a GitHub host.

The default authentication mode is a web-based browser flow. After completion, an
authentication token will be stored securely in the system credential store.
If a credential store is not found or there is an issue using it gh will fallback
to writing the token to a plain text file. See `gh auth status` for its
stored location.

Alternatively, use `--with-token` to pass in a token on standard input.
The minimum required scopes for the token are: `repo`, `read:org`, and `gist`.

Alternatively, gh will use the authentication token found in environment variables.
This method is most suitable for "headless" use of gh such as in automation. See
`gh help environment` for more info.

```

denesb commented on scripts/pull_github_pr.sh:
```diff
@@ -85,14 +70,18 @@ if [[ $nr_commits == 1 ]]; then
+	gh pr edit "${PR_NUM}" -b "${message}${closes}"
+	gh pr checkout "${PR_NUM}"
+	git rebase next
+	git push -f
```
Would be better to fully qualify this push, i.e. provide the remote branch name. Are we pushing to `next` or the PR branch here?

yaronkaikov replied:
The push is to the PR branch, I will make it more explicit

yaronkaikov replied:
Fixed

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8224/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8224/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9162/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1141/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8224/)
#### Build Details:
- Duration: 2 hr 16 min
- Builder: spider7.cloudius-systems.com

denesb commented:
> @denesb There is another solution instead of the PR (I just though about it). We have today an automated process in which we are adding a `promoted-to-master` label when the commit passes gating, Since we already know the PR number we can also close this PR during this process. This way we can simplify the script only doing merge for a single commit PR, and avoid adding the `closes` prefix
> 
> WDYT?

I am fine with that, but it would be useful to still keep a back-reference from the commit to the PR. It helps in git archeology. So maybe we can replace `Closes: #NNN` with some other keyword, which is not recognized by github, but which still communicates the same to humans. Github recognizes references even if there is no registered keyword anyway.

How would this solve the problem BTW? If we do the `git commit amend` I mentioned above, it would still change the SHA, right? In fact, even just the `git cherry-pick` already changes it, because the SHA includes the SAH of the ancestors -- and therefore it depends on the exact history. The only way to not change the SHA is merges (both fast-forward and non-fast-forward).

yaronkaikov commented:
> > @denesb There is another solution instead of the PR (I just though about it). We have today an automated process in which we are adding a `promoted-to-master` label when the commit passes gating, Since we already know the PR number we can also close this PR during this process. This way we can simplify the script only doing merge for a single commit PR, and avoid adding the `closes` prefix
> > WDYT?
> 
> I am fine with that, but it would be useful to still keep a back-reference from the commit to the PR. It helps in git archeology. So maybe we can replace `Closes: #NNN` with some other keyword, which is not recognized by github, but which still communicates the same to humans. Github recognizes references even if there is no registered keyword anyway.
> 
> How would this solve the problem BTW? If we do the `git commit amend` I mentioned above, it would still change the SHA, right? In fact, even just the `git cherry-pick` already changes it, because the SHA includes the SAH of the ancestors -- and therefore it depends on the exact history. The only way to not change the SHA is merges (both fast-forward and non-fast-forward).

After doing the `git commit amend` we also rebase and push the same commit to the PR. This is what we don't have today.

denesb commented:
> After doing the `git commit amend` we also rebase and push the same commit to the PR. This is what we don't have today.

How will this work with permissions? With a PR, the to-be-merged branch usually comes from a fork. I don't know if the maintainer who executes this command has write access to the fork as well. To test this, can you open a dummy PR and I will try to merge it with this script (to a dummy next). I cannot test with a PR of mine, as I have write access to my own fork of course.

yaronkaikov commented:
> > After doing the `git commit amend` we also rebase and push the same commit to the PR. This is what we don't have today.
> 
> How will this work with permissions? With a PR, the to-be-merged branch usually comes from a fork. I don't know if the maintainer who executes this command has write access to the fork as well. To test this, can you open a dummy PR and I will try to merge it with this script (to a dummy next). I cannot test with a PR of mine, as I have write access to my own fork of course.

https://github.com/scylladb/scylladb/pull/18393

Let's try on this one

yaronkaikov commented:
v2:
- added remote URL for push command
- after update move back to next branch

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8244/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8244/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1152/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9173/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8244/)
#### Build Details:
- Duration: 2 hr 21 min
- Builder: spider7.cloudius-systems.com

nyh commented:
> I am fine with that, but it would be useful to still keep a back-reference from the commit to the PR. It helps in git archeology. So maybe we can replace `Closes: #NNN` with some other keyword, which is not recognized by github, but which still communicates the same to humans. Github recognizes references even if there is no registered keyword anyway.

I agree. I often use this "Close #..." in a commit message to find the PR (otherwise, it's not obvious how to find).
It makes a lot of sense - and in some respects will be even better - "Closes #..." was always weird (why "closes") - a phrase like "From pull request: #...." (or something similar) would be good.

 
> How would this solve the problem BTW? If we do the `git commit amend` I mentioned above, it would still change the SHA, right? In fact, even just the `git cherry-pick` already changes it, because the SHA includes the SAH of the ancestors -- and therefore it depends on the exact history. The only way to not change the SHA is merges (both fast-forward and non-fast-forward).

Exactly what I thought :-( Since rebases change commit hashes, the commit message (which has nothing to do with the "Closes") is wrong.

yaronkaikov commented:
v2:

- simplify the solution using `git notes` to add reference of pull request so it will not update the commit hash
- PR description was updated

nyh commented:
> After doing the `git commit amend` we also rebase and push the same commit to the PR. This is what we don't have today.

I still don't understand...  :-(
1. You can't push "to the PR" - normally the PR comes from some repository I don't have write permissions for. I have write permissions to the Scylla "next" branch - not to the PR branch! I see @denesb asked the same question before, and I don't see an answer.
2. It's not clear to me how adding "Closes:" or not adding "Closes" changes **anything** here. In any case, the problem is that we want a single-patch commit to be rebased, not merged, and this is what causes the modification of the commit hash. Unless I'm missing something?
3. If you don't use "Closes:", how will github know to close the issue when the patch is promoted from next to master? Remember that we do a rebase, and moreover, maintainers may rebase the next branch again later (e.g., to dequeue faulty patches), so github will not recognize the patches as being merges (they aren't) and won't close the issue without the "Closes:".

All in all, it seems like mergify is buggy and needs to be fixed, I don't understand why we need to jump through hoops to fit Mergify's bugs, and still don't understand what it gives us that a 50-line script doesn't. I'm not volunteering to write this script (and I'm very far from being a github automation expert) - but it seems we're investing a lot of effort into fitting mergify's demands (and still not getting it right...), so I wonder if wouldn't have been easier just not to use mergify and write exactly the code we need.

yaronkaikov commented:
> > After doing the `git commit amend` we also rebase and push the same commit to the PR. This is what we don't have today.
> 
> I still don't understand... :-(
> 
> 1. You can't push "to the PR" - normally the PR comes from some repository I don't have write permissions for. I have write permissions to the Scylla "next" branch - not to the PR branch! I see @denesb asked the same question before, and I don't see an answer.

All dev team member in Github have `triage` permissions, meaning they can't edit branches other than on their fork, this PR is not about that. Moving the dev team to have to write permission with a specific ruleset that allows editing only `mergify/**` branches is a possible solution.

> 2. It's not clear to me how adding "Closes:" or not adding "Closes" changes **anything** here. In any case, the problem is that we want a single-patch commit to be rebased, not merged, which causes the modification of the commit hash. Unless I'm missing something?

Today when we added the `closes` prefix we changed the commit SHA, but the PR (in Github) is not updated. IIUC there are 2 reasons for adding the `closes` prefix:
1. We want to close the PR rather than merge it (for single commit PR), and this is how we can do it
2. Have some reference from the commit to the PR

The change in this PR (v2), is solving point 2 without changing the commit SHA

> 3. If you don't use "Closes:", how will github know to close the issue when the patch is promoted from next to master? Remember that we do a rebase, and moreover, maintainers may rebase the next branch again later (e.g., to dequeue faulty patches), so github will not recognize the patches as being merges (they aren't) and won't close the issue without the "Closes:".

So I missed some important changes in my PR :-(, The part where we close the PR once the patch is promoted. I will add it soon, in the meantime, switching back to DRAFT. 

> 
> All in all, it seems like mergify is buggy and needs to be fixed, I don't understand why we need to jump through hoops to fit Mergify's bugs, and still don't understand what it gives us that a 50-line script doesn't. I'm not volunteering to write this script (and I'm very far from being a github automation expert) - but it seems we're investing a lot of effort into fitting mergify's demands (and still not getting it right...), so I wonder if wouldn't have been easier just not to use mergify and write exactly the code we need.

I am not aware of any new tool that was implemented without having some issues at the beginning, this is what the integration is all about. I think that at least some of the issues we are facing is related to the way we do things in Scylla. In Scylla-pkg for example (we also using Mergify), things are more smoothly. 
I am not saying it's the perfect tool but IMO the current problems we are facing can be resolved

nyh commented:
> All dev team member in Github have `triage` permissions, meaning they can't edit branches other than on their fork, this PR is not about that. Moving the dev team to have to write permission with a specific ruleset that allows editing only `mergify/**` branches is a possible solution.

Here is a random PR: https://github.com/scylladb/scylladb/pull/18497. The pull request's branch belongs to the user patjed41 - it's not on Scylla's repository, or a "mergify/" branch. I guess I don't understand what you're proposing - I don't see in your patch any code which tries to write back the PR branch, so I am not sure you even said it was necessary. But I have a guess what maybe you meant - more on that below.

> 
> > 2. It's not clear to me how adding "Closes:" or not adding "Closes" changes **anything** here. In any case, the problem is that we want a single-patch commit to be rebased, not merged, which causes the modification of the commit hash. Unless I'm missing something?
> 
> Today when we added the `closes` prefix we changed the commit SHA, but the PR (in Github) is not updated.

But what I'm trying to say is that "changing the commit SHA" doesn't happen because of the "closes prefix" - we always add this "Closes" thing, not just for single-commit PRs. The special thing about single-commit PRs is not the closes line - it's the fact we want (and our script does it for us) to _rebase_ the single commit on top of the current next - not to "merge". It is this rebase which causes the commit hash to change. And as I noted, it can even change again later! If a maintainer needs to dequeue a patch, they rebase next again, and commit hashes changes yet again!

What all of this means is that:
"git cherry-pick -x" needs to be done after a patch has reached a stable commit hash - i.e., master,  if we want it to print the right commit hash.

This is how we always did things before the advent of this automation - we only did the backport after the patch reached master, and we did the cherry-pick *from* master, so it has permanent, master, commit hashes.

"mergify" needs to know to do this. It can also use the wrong commit hash reference if it promises to change it later before the backport PR is finally merged. Maybe this is what you meant by modifying the pull request? If maintainers had permissions to write to the mergify/ branch, and if our script knew how to rewrite cherry-pick commit hashes it finds in the backported patch, we could "fix" the reference before merging the backport. But this is not what this current patch seems to do.

>  IIUC there are 2 reasons for adding the `closes` prefix:
> 
>     1. We want to close the PR rather than merge it (for single commit PR), and this is how we can do it
> 
>     2. Have some reference from the commit to the PR
> 
> 
> The change in this PR (v2), is solving point 2 without changing the commit SHA

This doesn't sound true - both because of the "rebase" thing I mentioned earlier, but also - why does the commit SHA change or not change depending on whether the text has "Closes: 123" or another text like "Based on PR 123"?

> I am not aware of any new tool that was implemented without having some issues at the beginning, this is what the integration is all about. I think that at least some of the issues we are facing is related to the way we do things in Scylla. In Scylla-pkg for example (we also using Mergify), things are more smoothly. I am not saying it's the perfect tool but IMO the current problems we are facing can be resolved

It's always easier to accept a tool when you have no prior expectations on what it needs to do. But the Scylla developers already have some expectations (originally set by Avi based on his experiences in the Linux project), on git "hygiene": 
1. They expect single patches to not look like a merge of a single-patch branch, but just as a single patch. This is the source of the problems discussed in this issue (it's not the "Closes:", unless I'm completely misunderstanding something).
2. They don't want junk like references to incorrect commit hashes, in their "git log".

yaronkaikov commented:
@nyh I will schedule a meeting on Sunday so we can talk about it. I think it will be much more affective

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8481/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8481/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9356/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1332/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8481/)
#### Build Details:
- Duration: 2 hr 16 min
- Builder: spider6.cloudius-systems.com

denesb commented:
> Maybe this is what you meant by modifying the pull request? If maintainers had permissions to write to the mergify/ branch, and if our script knew how to rewrite cherry-pick commit hashes it finds in the backported patch, we could "fix" the reference before merging the backport. But this is not what this current patch seems to do.

I think this would be the path of least resistance. Make `pull_gh_pr.sh` detect bad "cherry picked from" hashes and correct them, just prior to applying the commit to `next`. This will work in all cases. As @nyh already noted, the approach taken in this PR will break if `next` is rebased and we do rebase `next` from time-to-time.

yaronkaikov commented:
@denesb @nyh Just an update on this one. The guys from Mergify found a bug related to pointing in the PR body to the wrong commit SHA. They are working on a fix, and it should be ready probably early next week, I am keeping this PR open for now, and once they have the fix, the solution will be some small change in the mergify conf file.

yaronkaikov commented:
> > After doing the `git commit amend` we also rebase and push the same commit to the PR. This is what we don't have today.
> 
> I still don't understand... :-(
> 
> 1. You can't push "to the PR" - normally the PR comes from some repository I don't have write permissions for. I have write permissions to the Scylla "next" branch - not to the PR branch! I see @denesb asked the same question before, and I don't see an answer.

I am working on a fix on this one as well, hope to have some progress by tomorrow :-)

yaronkaikov commented:
@nyh @denesb Mergify team fixed the cherry-pick ref to wrong commit :-), you can see a working example in https://github.com/scylladb/scylladb/pull/18724 

I am closing this PR, no change from our side is needed
